fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java

```diff
@@ -214,6 +228,161 @@ public TGetDbsResult getDbNames(TGetDbsParams params) throws TException {
         return result;
     }
 
+    private static ColumnDef initColumnfromThrift(TColumnDesc tColumnDesc, String comment) {
+        TypeDef typeDef = TypeDef.createTypeDef(tColumnDesc);
+        boolean isAllowNull = tColumnDesc.isIsAllowNull();
+        ColumnDef.DefaultValue defaultVal = ColumnDef.DefaultValue.NOT_SET;
+        // Dynamic table's Array default value should be '[]'
+        if (typeDef.getType().isArrayType()) {
+            defaultVal = ColumnDef.DefaultValue.ARRAY_EMPTY_DEFAULT_VALUE;
+        }
+        return new ColumnDef(tColumnDesc.getColumnName(), typeDef, false, null, isAllowNull, defaultVal,
+                    comment, true);
+    }
+
+    @Override
+    public TAddColumnsResult addColumns(TAddColumnsRequest request) throws TException {
+        String clientAddr = getClientAddrAsString();
+        LOG.debug("schema change clientAddr: {}, request: {}", clientAddr, request);
+
+        TStatus status = new TStatus(TStatusCode.OK);
+        List<TColumn> allColumns = new ArrayList<TColumn>();
+
+        Env env = Env.getCurrentEnv();
+        InternalCatalog catalog = env.getInternalCatalog();
+        int schemaVersion = 0;
+        try {
+            if (!env.isMaster()) {
+                status.setStatusCode(TStatusCode.ILLEGAL_STATE);
+                status.addToErrorMsgs("retry rpc request to master.");
+                TAddColumnsResult result = new TAddColumnsResult(status, request.getTableId(), allColumns, 0);
+                LOG.debug("result: {}", result);
+                return result;
+            }
+            TableName tableName = new TableName("", request.getDbName(), request.getTableName());
+            if (request.getTableId() > 0) {
+                tableName = catalog.getTableNameByTableId(request.getTableId());
+            }
+            if (tableName == null) {
+                throw new MetaNotFoundException("table_id " + request.getTableId() + " does not exist");
+            }
+
+            Database db = catalog.getDbNullable(tableName.getDb());
+            if (db == null) {
+                throw new MetaNotFoundException("db " + tableName.getDb() + " does not exist");
+            }
+
+            List<TColumnDef> addColumns = request.getAddColumns();
+            boolean queryMode = false;
+            if (addColumns == null || addColumns.size() == 0) {
+                queryMode = true;
+            }
+
+            // rpc only olap table
+            OlapTable olapTable = (OlapTable) db.getTableOrMetaException(tableName.getTbl(), TableType.OLAP);
+            olapTable.writeLockOrMetaException();
+
+            try {
+                List<ColumnDef> columnDefs = new ArrayList<ColumnDef>();
+
+                // prepare columnDefs
+                for (TColumnDef tColumnDef : addColumns) {
+                    if (request.isTypeConflictFree()) {
```

Comment:
The name 'isTypeConflictFree' is not so intuitive. 'allowTypeConflict' may be better.

Reply:
ok